cmd/opampsupervisor/specification/README.md

```diff
@@ -220,19 +225,69 @@ telemetry:
+the configuration files specified are applied in the order they are specified.
+In other words, configuration files are merged from the top of the list to the bottom.
+Configuration added by files at the top of the list may be overwritten by the later ones.
+
+The indicated configuration files are merged in memory and the resulting configuration
+is written to `<storage::directory>/effective.yaml`.
 
-Please be aware that when using the `.agent.config_files` parameter,
-the configuration files specified are applied after the configuration from the OpAMP server.
-After the configuration files, arguments present in `.agent.args` are passed to the executable binary.
-The environmanet variables specified in `.agent.env` are set in the collector process environment.
+There are a few "magic" configuration files that can be used to completely
+customize the priority order of the Collector configuration. Below are the available
+values and what they represent:
 
-The following configuration:
+- `$OPAMP_EXTENSION_CONFIG`: configuration for the OpAMP extension to connect to the Supervisor.
+- `$OWN_METRICS_CONFIG`: configuration for the Collector to report its own metrics.
+- `$BUILTIN_CONFIG`: built-in base configuration.
+- `$REMOTE_CONFIG`: remote configuration received by the Supervisor.
+
+If **none** of these are explicitly specified, they are automatically prepended to the
+list of configuration files. This ensures that user-provided configuration files have
+higher priority.
+
+These magic files can be mixed with user-provided configuration files to create complex
+configuration merge orders, for instance, creating base-layer configuration at the
+lowest priority while keeping compliance configuration at the highest priority:
+
+```yaml
+agent:
+  config_files:
+    - base_config.yaml
+    - $OPAMP_EXTENSION_CONFIG
+    - $OWN_METRICS_CONFIG
+    - $BUILTIN_CONFIG
+    - $REMOTE_CONFIG
+    - compliance_config.yaml
+```
+
+**VERY IMPORTANT**: if a single magic file is found in `agent::config_files`,
+the Supervisor will not automatically add the others. The user is responsible for ensuring
+that all of them are present. Missing a magic file can cause their respective feature to
+not work correctly, potentially affecting the Supervisor's and Collector's behavior.
+The list below contains a few, but not all, examples of what can happen if a gven
+magic file is missing:
+
+- If the `$OPAMP_EXTENSION_CONFIG` is not specified, the Collector
+  will not be able to connect to the Supervisor, triggering a restart loop.
+- If the `$OWN_METRICS_CONFIG` is not specified, the Collector will not be able
+  to report its own metrics.
+- If the `$BUILTIN_CONFIG` is not specified, the Collector will not have the
+  expected identifying, non-identifying and resource attributes.
+- If the `$REMOTE_CONFIG` is not specified, the Collector will not be able to
+  apply the remote configuration received from the Supervisor.
```

Comment:
Do we need to provide validation against the merged config that the things we expect to be present are present? I feel like we are opening up ourselves to a lot of troubleshooting issues from users holding this feature incorrectly

Reply:
@TylerHelmuth I thought about this a bit. One way we could avoid this problem is to automatically prepend any non-specified special file, so this way what we expect to be there will be there **unless** modified by later/higher-level configurations at the expense of having a bit of "hidden" behavior. I think this can be a good tradeoff. WDYT, @TylerHelmuth?

Reply:
Ya always prepending one of our special configurations if it is not specified in the list by the user feels like a safer default.

Reply:
I'm okay taking this on a case-by-case basis: for example, I'm not as worried about the own metrics config as the other ones.

However in general I think we should enable default anything that would cause the Supervisor to break in most cases (removing the OpAMP extension config) or introduce unclear behavior (the remote configuration capability is enabled, but the remote config merge order isn't set, so don't use remote config). For now I would be okay just always including all of the special configurations.